README.md
# Тестовое задание API Рекламы

## Настройка окружения
### Docker
Развернуть приложение:
~~~
docker-compose up --build
~~~

Применить миграции:
~~~
docker-compose exec app php ./scripts/migrate-prod
docker-compose exec app php ./scripts/migrate-test
~~~

Добавить в файл <Директория Windows>\System32\drivers\etc\hosts связь IP-адреса с именем хоста:
~~~
127.0.0.1    ads-api.loc
~~~

### Vagrant
Для автоматизации настройки рабочего окружения используется [Vagrant](https://www.vagrantup.com/):
~~~
vagrant up
~~~

Добавить в файл <Директория Windows>\System32\drivers\etc\hosts связь IP-адреса с именем хоста:
~~~
127.0.0.1    ads-api.loc
~~~

Основа - образ [Ubuntu 18.04](https://app.vagrantup.com/bento/boxes/ubuntu-18.04/versions/202112.19.0). В случае, если не удаётся автоматизировано загрузить образ (box) из репозитория, можно добавить его вручную ([файл образа](https://app.vagrantup.com/bento/boxes/ubuntu-18.04/versions/202112.19.0/providers/virtualbox.box)):
~~~
vagrant box add "bento/ubuntu-18.04" ./cf00babe-4123-4902-ad3a-c98f80de0e30
~~~

### Windows (без Vagrant/Docker)

* Apache 2.4
* PHP 7.2.25
* MySQL для работы, SQLite3 для тестов

#### Apache

Настройка виртуального хоста:

1. Подключить конфиг виртуальных хостов к основному конфигу веб-сервера. Добавить в конфиг веб-сервера (<Директория Apache>/conf/httpd.conf):
~~~
# Virtual hosts
Include conf/extra/httpd-vhosts.conf
~~~

2. Скопировать [конфиг](https://github.com/cosmastar112/ads_api/blob/master/apache/ads-api.loc.conf) виртуального хоста ads-api.loc в директорию (которую нужно создать) с виртуальными конфигами - <Директория Apache>/conf/extra/vh.

3. Подключить конфиг виртуального хоста ads-api.loc в конфиге виртуальных хостов. Добавить в <Директория Apache>/conf/extra/httpd-vhosts.conf:
~~~
Include conf/extra/vh/ads-api.loc.conf
~~~

4. Создать ссылку на директорию web проекта в <Директория Apache>/htdocs. Воспользоваться утилитой [junction](https://github.com/cosmastar112/ads_api/blob/master/apache/junction64.exe):
~~~
<Директория проекта>\apache\junction64.exe <Директория Apache>\htdocs\ads-api.loc <Директория проекта>\web
~~~

5. Добавить в файл <Директория Windows>\System32\drivers\etc\hosts связь IP-адреса с именем хоста:
~~~
127.0.0.1    ads-api.loc
127.0.0.1    db
~~~

#### Структура БД

##### Рабочая БД (MySQL)
Создать БД
~~~
CREATE SCHEMA `ads` ;
~~~
Создать пользователя "ad-api" (в БД "ads"), выдать ему права на БД "ads"
~~~
CREATE USER 'ad-api'@'localhost' IDENTIFIED BY 'password';
GRANT ALL ON ads.* TO 'ad-api'@'localhost';
~~~
Применить миграции для БД "ads"
~~~
cd /d <Директория проекта>/scripts
php migrate-prod
~~~
или
~~~
<Директория проекта>/scripts/migrate-prod.bat
подтвердить запуск миграций (WARNING библиотеки миграций)
~~~

##### Тестовая БД (SQLite)
SQLite 3.37.2 2022-01-06 13:25:41

Создать БД и применить миграции для "ads-test"
~~~
cd /d <Директория проекта>/scripts
php migrate-test
~~~
или
~~~
<Директория проекта>/scripts/migrate-test.bat
~~~
В результате в директории db появится файл БД ads-test.db. На существование файла БД [полагается модуль DB](https://github.com/cosmastar112/ads_api/blob/master/codeception.yml#L13) тестового фреймворка codeception.
Для интерактивного взаимодействия с БД можно использовать [cmd-утилиту](https://github.com/cosmastar112/ads_api/blob/master/db/bin/sqlite3.exe).

## Использование (на примере curl)

### Создание
~~~
curl -X POST http://ads-api.loc:8080/index.php/ads -d "text=Advertisement1&price=300&limit=1000&banner=https://linktoimage.png"
~~~

### Редактирование
~~~
curl -X POST http://ads-api.loc:8080/index.php/ads/1 -d "text=Advertisement2&price=500&limit=800&banner=https://anotherlinktoimage.png"
~~~

### Открутка
~~~
curl http://ads-api.loc:8080/index.php/ads/relevant
~~~

## Компоненты проекта
###	Миграции
Используется библиотека миграций doctrine/migrations, конфигурируемая с помощью файлов:
* [migrations.php](https://github.com/cosmastar112/ads_api/blob/master/config/migrations.php)
* [migrations-db.php](https://github.com/cosmastar112/ads_api/blob/master/config/migrations-db.php)
* [migrations-db-test.php](https://github.com/cosmastar112/ads_api/blob/master/config/migrations-db-test.php)

Миграции находятся в директории [migrations](https://github.com/cosmastar112/ads_api/tree/master/migrations).

### Документация

Для создания документации используется библиотека [phpDocumentor](https://docs.phpdoc.org/3.0/guide/getting-started/installing.html) v3.3.0. Запуск:
~~~
Windows:
cd /d <Директория проекта>
php ./vendor/bin/phpDocumentor.phar <настройки>

или

<Директория проекта>/scripts/doc.bat

или

cd /d <Директория проекта>/scripts
php doc

Linux:
cd /d <Директория проекта>
php ./vendor/bin/phpDocumentor.phar <настройки>

или

cd <Директория проекта>/scripts
php doc
~~~
Результатом является HTML-документация в директории [docs/phpdoc](https://github.com/cosmastar112/ads_api/tree/master/docs/phpdoc).

#### Статический анализатор

Используется статический анализатор [Psalm](https://psalm.dev/). Запуск:
~~~
Windows:
cd /d <Директория проекта>
./vendor/bin/psalm <настройки>

или

cd /d <Директория проекта>/scripts
run-static-analysis-tool.bat

Linux:
cd /d <Директория проекта>
./vendor/bin/psalm <настройки>

или

cd <Директория проекта>/scripts
run-static-analysis-tool.sh
~~~

#### Unit-тесты

Используется [Codeception](https://codeception.com/). Запуск:
~~~
Windows:
cd /d <Директория проекта>
./vendor/codeception/codeception/codecept run unit

или

<Директория проекта>/scripts/doc.bat

Linux:
cd /d <Директория проекта>
./vendor/codeception/codeception/codecept run unit

или

cd <Директория проекта>/scripts
run-static-analysis-tool.sh
~~~

## Описание проекта
API состоит из элементов:
* создание
* редактирование
* открутка

| Название элемента	| Маршрут	| Тип запроса	| Примечание |
| --- | --- | --- | --- |
| создание | /ads |	POST | |
| редактирование | /ads/\<number\> | POST | В строке запроса передается id редактируемого элемента |
| открутка | /ads/relevant | GET |

### Создание
Пример запроса
~~~
POST /ads HTTP/1.1
Host: localhost
Content-Type: application/x-www-form-urlencoded
...

text=Advertisement1&price=300&limit=1000&banner=https://linktoimage.png
~~~

Описание полей
| Поле | Описание |	Тип |
| --- | --- | --- |
| text | Заголовок объявления | Строковый |
| price |	Стоимость одного показа |	Числовой |
| limit	| Лимит показов	| Числовой |
| banner | Ссылка на картинку | Строковый |

Пример ответа
~~~
HTTP/1.1 200 OK
Content-Type: application/json
...

{
  "message": "OK",
  "code": 200,
  "data": {
     "id": 123,
     "text": "Advertisement1",
     "banner": "https://linktoimage.png"   
  }
}
~~~

Пример ответа с ошибкой валидации поля
~~~
HTTP/1.1 200 OK
Content-Type: application/json
...

{
  "message": "Invalid banner link",
  "code": 400,
  "data": {}
}
~~~

### Редактирование
Пример запроса
~~~
POST /ads/1234 HTTP/1.1
Host: localhost
Content-Type: application/x-www-form-urlencoded
...

text=Advertisement123&price=450&limit=1200&banner=https://linktoimage.png
~~~

Пример ответа
~~~
HTTP/1.1 200 OK
Content-Type: application/json
...

{
  "message": "OK",
  "code": 200,
  "data": {
    "id": 123,
    "text": "Advertisement123",
    "banner": "https://linktoimage.png"
  }
}
~~~

### Открутка
Возвращается id, text и banner объявления. Выбирается по следующим условиям:
* У него должна быть самая высокая цена
* Количество открученных показов этого объявления не должно превышать лимит показов (поле limit).

Пример запроса
~~~
GET /ads/relevant HTTP/1.1
Host: localhost
...

~~~
Пример ответа
~~~
HTTP/1.1 200 OK
Content-Type: application/json
...

{
  "message": "OK",
  "code": 200,
  "data": {
     "id": 123,
     "text": "Advertisement1",
     "banner": "https://linktoimage.png"   
  }
}
~~~

## Описание реализации

Входной [скрипт](https://github.com/cosmastar112/ads_api/blob/master/web/index.php) расположен в директории [web](https://github.com/cosmastar112/ads_api/tree/master/web).
В нём создаётся экземпляр [приложения](https://github.com/cosmastar112/ads_api/blob/master/app/Application.php), который управляет обработкой запроса. 
Для настройки приложения используются конфиг-файлы, расположенные в директории [config](https://github.com/cosmastar112/ads_api/tree/master/config):
* [router.php](https://github.com/cosmastar112/ads_api/blob/master/config/router.php) для [Роутера](https://github.com/cosmastar112/ads_api/blob/master/app/Router.php)
* [db-mysql.php](https://github.com/cosmastar112/ads_api/blob/master/config/db-mysql.php) (или [db-sqlite.php](https://github.com/cosmastar112/ads_api/blob/master/config/db-sqlite.php)) для объекта, который отвечает за [соединение с БД](https://github.com/cosmastar112/ads_api/blob/master/app/Db.php)

Для обработки запроса Приложение использует Роутер. Задача роутера – определить, существует ли реальный маршрут путём сопоставления запрашиваемого маршрута с доступными маршрутами, перечисленными в конфиге роутера. Для маршрутизации используется библиотека [FastRoute](https://github.com/nikic/FastRoute).

Если маршрут определить не удалось, будет возвращена [ошибка 404](https://github.com/cosmastar112/ads_api/blob/master/errors/404.html).
Если метод не поддерживается, будет возвращена [ошибка 405](https://github.com/cosmastar112/ads_api/blob/master/errors/405.html).
Если маршрут существует и указан верный HTTP-метод, создается объект контроллера и вызывается нужный метод (экшен).

Все контроллеры располагаются в директории [controllers](https://github.com/cosmastar112/ads_api/tree/master/controllers) и должны наследоваться от [базового класса](https://github.com/cosmastar112/ads_api/blob/master/app/Controller.php). Приложение возвращает результат работы метода объекта.

В экшене (создание, редактирование, открутка), в зависимости от элемента, контроллера осуществляется:
* доступ к параметрам запроса (если они есть)
* валидация параметров (если они есть)
* выполнение операции
*	возврат результата.

Для валидации используется [модель](https://github.com/cosmastar112/ads_api/blob/master/models/Ad.php), которая располагается в директории [models](https://github.com/cosmastar112/ads_api/tree/master/models).
В случае ошибки валидации будет возвращена ошибка 400.
Для выполнения операции используется [другой объект](https://github.com/cosmastar112/ads_api/blob/master/models/rep/AdRep.php), в котором инкапсулирована работа с БД (паттерн Репозиторий). Для этого объект реализует [интерфейс](https://github.com/cosmastar112/ads_api/blob/master/app/IRepository.php).
Для работы с БД используется объект [Db](https://github.com/cosmastar112/ads_api/blob/master/app/Db.php) – обёртка над [PDO](https://www.php.net/manual/en/book.pdo.php). Обёртка обеспечивает работу с СУБД MySQL и SQLite. Соединение с БД открывается один раз при создании Приложения; доступ к нему предоставляется с помощью метода getDb.
В свою очередь, для использования приложения предназначен статический метод Application::getApp(), который возвращает ссылку на экземпляр Приложения, которая сохраняется при его создании.

## Развитие (TODO)
* Использовать библиотеки для решения задачи (с помощью composer):
  *	ORM для работы с БД
  *	Привести код к стандарту PSR-12
* Написать тесты (как минимум, модульные)
*	Новый алгоритм открутки


CHANGELOG.md
# История изменений Тестовое задание API Рекламы (ads_api)
## 0.8.0 (В разработке)
* В этом релизе пока нет изменений

## 0.7.0 (26 Июня 2022)
* Enc [#22](https://github.com/cosmastar112/ads_api/issues/22): Удаление сгенерированной документации
* Enc [#20](https://github.com/cosmastar112/ads_api/issues/20): Использование Docker для автоматизации настройки окружения для разработки
* Enc [#24](https://github.com/cosmastar112/ads_api/issues/24): Документация: настройка виртуального хоста Apache

## 0.6.0 (03 Мая 2022)
* Enc [#21](https://github.com/cosmastar112/ads_api/issues/21): Использование Vagrant для автоматизации настройки окружения для разработки
* Enc [#14](https://github.com/cosmastar112/ads_api/issues/14): Документация: команда запуска тестов

## 0.5.0 (27 Марта 2022)
* Enc [#17](https://github.com/cosmastar112/ads_api/issues/17): Использование статического анализатора (Psalm)
* Enc [#15](https://github.com/cosmastar112/ads_api/issues/15): Документирование кода (phpDocumentor)

## 0.4.0 (19-03-2022)
* Enc [#7](https://github.com/cosmastar112/ads_api/issues/7): Документация: скрипт создания пользователя БД (MySQL)
* Enc [#6](https://github.com/cosmastar112/ads_api/issues/6): Документация: команда запуска миграций
* Enc [#5](https://github.com/cosmastar112/ads_api/issues/5): Использование БД SQLite для тестирования (также возможно использование в качестве рабочей БД); unit-тесты AdRep

## 23-02-2022
* Enc [#10](https://github.com/cosmastar112/ads_api/issues/10): Использование тестового фреймворка Codeception, unit-тесты models\Ad

## 20-02-2022
* Enc [#8](https://github.com/cosmastar112/ads_api/issues/8): Использование механизма PSR-4 атозагрузки классов composer

## 02-11-2021
* Enc [#3](https://github.com/cosmastar112/ads_api/issues/3): Для маршрутизации используется библиотека nikic/fast-route

## 29-10-2021
* Fix [#1](https://github.com/cosmastar112/ads_api/issues/1): Для описания структуры БД ипользуется библиотека doctrine/migrations

## 18-10-2021
* Релиз первой версии

controllers/Ads.php
<?php

namespace controllers;

use app\Controller;
use app\Application;
use models\Ad;
use models\rep\AdRep;

/**
 * Контроллер рекламы (объявлений)
*/
class Ads extends Controller
{
    /**
     * @internal Экземпляр БД.
     * @var \PDO|null
     * @link https://www.php.net/manual/ru/book.pdo.php
     */
    private $_db = null;

    /**
     * @return \PDO Экземпляр БД.
     */
    public function getDb()
    {
        if (!($this->_db instanceof \PDO)) {
            $this->_db = Application::getApp()->getDb();
        }
        return $this->_db;
    }

    /**
     * Создать рекламу (объявление)
     * @return string
     * @throws \Exception Если возникла ошибка при подготовке ответа или сохранении данных.
     */
    public function create()
    {
        header('Content-Type: application/json');

        /** @var string|null Описание. */
        $text = $this->getPostBodyParam('text');
        /** @var string|null Цена объявления. */
        $price = $this->getPostBodyParam('price');
        /** @var string|null Кол-во показов. */
        $limit = $this->getPostBodyParam('limit');
        /** @var string|null URL-адрес баннера (картинки). */
        $banner = $this->getPostBodyParam('banner');

        /** @var \models\Ad Модель. */
        $model = new Ad(null, $text, $price, $limit, $banner);
        //валидация
        if (!$model->validate('create')) {
            /** @var null|string Первая ошибка валидации */
            $firstError = $model->getFirstError();
            $response = [
                'message' => $firstError,
                'code' => 400,
                'data' => [],
            ];

            /** @var string|false Закодированная JSON-строка. */
            $result = json_encode($response);
            if ($result === false) {
                throw new \Exception("Ошибка при подготовке ответа");
            }

            return $result;
        }

        /** @var \models\rep\AdRep Репозиторий рекламы. */
        $rep = new AdRep($this->getDb());
        if ($rep->save($model)) {
            /** @var string|false Закодированная JSON-строка. */
            $result = json_encode([
                'message' => 'OK',
                'code' => 200,
                'data' => [
                    'text' => $text,
                    'price' => $price,
                    'limit' => $limit,
                    'banner' => $banner,
                ]
            ]);
            if ($result === false) {
                throw new \Exception("Ошибка при подготовке ответа");
            }

            return $result;
        } else {
            throw new \Exception("Ошибка при сохранении данных.");
        }
    }

    /**
     * Изменить рекламу (объявление)
     * @return string
     * @throws \Exception Если возникла ошибка при подготовке ответа или сохранении данных.
     */
    public function update()
    {
        header('Content-Type: application/json');

        /** @var string|null Идентификатор. */
        $id = $this->getParam('id');
        /** @var string|null Описание. */
        $text = $this->getPostBodyParam('text');
        /** @var string|null Цена объявления. */
        $price = $this->getPostBodyParam('price');
        /** @var string|null Кол-во показов. */
        $limit = $this->getPostBodyParam('limit');
        /** @var string|null URL-адрес баннера (картинки). */
        $banner = $this->getPostBodyParam('banner');

        /** @var \models\Ad Модель. */
        $model = new Ad($id, $text, $price, $limit, $banner);
        //валидация
        if (!$model->validate('update')) {
            /** @var null|string Первая ошибка валидации */
            $firstError = $model->getFirstError();
            $response = [
                'message' => $firstError,
                'code' => 400,
                'data' => [],
            ];

            /** @var string|false Закодированная JSON-строка. */
            $result = json_encode($response);
            if ($result === false) {
                throw new \Exception("Ошибка при подготовке ответа");
            }

            return $result;
        }

        /** @var \models\rep\AdRep Репозиторий рекламы. */
        $rep = new AdRep($this->getDb());
        if ($rep->save($model)) {
            /** @var string|false Закодированная JSON-строка. */
            $result = json_encode([
                'message' => 'OK',
                'code' => 200,
                'data' => [
                    'id' => $id,
                    'text' => $text,
                    'price' => $price,
                    'limit' => $limit,
                    'banner' => $banner,
                ]
            ]);
            if ($result === false) {
                throw new \Exception("Ошибка при подготовке ответа");
            }

            return $result;
        } else {
            throw new \Exception("Ошибка при сохранении данных.");
        }
    }

    /**
     * Выбрать рекламу (объявление) для показа
     * @return string
     * @throws \Exception Если возникла ошибка при подготовке ответа или сохранении данных.
     */
    public function relevant()
    {
        header('Content-Type: application/json');

        /** @var \models\rep\AdRep Репозиторий рекламы. */
        $rep = new AdRep($this->getDb());
        /** @var array|false Выбранная для показа модель рекламы. */
        $model = $rep->getRelevant();
        if ($model) {
            /** @var string|false Закодированная JSON-строка. */
            $result = json_encode([
                'message' => 'OK',
                'code' => 200,
                'data' => [
                    'id' => $model['id'],
                    'text' => $model['text'],
                    'banner' => $model['banner'],
                ]
            ]);
            if ($result === false) {
                throw new \Exception("Ошибка при подготовке ответа");
            }

            return $result;
        } else {
            throw new \Exception("Ошибка при загрузке данных.");
        }
    }
}